File: docs/examples/java/RealTimeTransactionProcessing.java
import com.chain.api.*;
import com.chain.http.*;
import com.chain.signing.*;

class RealTimeTransactionProcessing {
  public static void main(String[] args) throws Exception {
    Context context = new Context();
    setup(context);

    // snippet processing-thread
    new Thread(() -> {
        processingLoop(context);
    }).start();
    // endsnippet

    // snippet issue
    Transaction.Template issuance = new Transaction.Builder()
      .addAction(new Transaction.Action.Issue()
        .setAssetAlias("gold")
        .setAmount(100)
      ).addAction(new Transaction.Action.ControlWithAccount()
        .setAccountAlias("alice")
        .setAssetAlias("gold")
        .setAmount(100)
      ).build(context);

    Transaction.submit(context, HsmSigner.sign(issuance));
    // endsnippet

    Thread.sleep(1000);

    // snippet transfer
    Transaction.Template transfer = new Transaction.Builder()
      .addAction(new Transaction.Action.SpendFromAccount()
        .setAccountAlias("alice")
        .setAssetAlias("gold")
        .setAmount(50)
      ).addAction(new Transaction.Action.ControlWithAccount()
        .setAccountAlias("bob")
        .setAssetAlias("gold")
        .setAmount(50)
      ).build(context);

    Transaction.submit(context, HsmSigner.sign(transfer));
    // endsnippet

    Thread.sleep(1000);
    System.exit(0);
  }

  public static void setup(Context context) throws Exception {
    MockHsm.Key key = MockHsm.Key.create(context);
    HsmSigner.addKey(key, MockHsm.getSignerContext(context));

    new Asset.Builder()
      .setAlias("gold")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(context);

    new Account.Builder()
      .setAlias("alice")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(context);

    new Account.Builder()
      .setAlias("bob")
      .addRootXpub(key.xpub)
      .setQuorum(1)
      .create(context);

    // snippet create-feed
    Transaction.Feed feed = Transaction.Feed.create(
      context,
      "local-transactions",
      "is_local='yes'"
    );
    // endsnippet
  }

  public static void processingLoop(Context context) {
    try {
      // snippet get-feed
      Transaction.Feed feed = Transaction.Feed.getByAlias(
        context,
        "local-transactions"
      );
      // endsnippet

      // snippet processing-loop
      while (true) {
        Transaction tx = feed.next(context);
        processTransaction(tx);
        feed.ack(context);
      }
      // endsnippet
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  // snippet processor-method
  public static void processTransaction(Transaction tx) {
    System.out.println("New transaction at " + tx.timestamp);
    System.out.println("\tID: " + tx.id);

    for (int i = 0; i < tx.inputs.size(); i++) {
      Transaction.Input input = tx.inputs.get(i);
      System.out.println("\tInput " + i);
      System.out.println("\t\tType: " + input.type);
      System.out.println("\t\tAsset: " + input.assetAlias);
      System.out.println("\t\tAmount: " + input.amount);
      System.out.println("\t\tAccount: " + input.accountAlias);
    }

    for (int i = 0; i < tx.outputs.size(); i++) {
      Transaction.Output output = tx.outputs.get(i);
      System.out.println("\tOutput " + i);
      System.out.println("\t\tType: " + output.type);
      System.out.println("\t\tPurpose: " + output.purpose);
      System.out.println("\t\tAsset: " + output.assetAlias);
      System.out.println("\t\tAmount: " + output.amount);
      System.out.println("\t\tAccount: " + output.accountAlias);
    }
  }
  // endsnippet
}

File: docs/1.1/core/examples/node/realTimeTransactionProcessing.js
const chain = require('chain-sdk')

const client = new chain.Client()
const signer = new chain.HsmSigner()
let key

// snippet processor-method
const processTransaction = tx => {
  console.log(`New transaction at ${tx.timestamp}`)
  console.log(`\tID: ${tx.id}`)

  tx.inputs.forEach((input, index) => {
    console.log(`\tInput ${index}`)
    console.log(`\t\tType: ${input.type}`)
    console.log(`\t\tAsset: ${input.assetAlias}`)
    console.log(`\t\tAmount: ${input.amount}`)
    console.log(`\t\tAccount: ${input.accountAlias}`)
  })

  tx.outputs.forEach((output, index) => {
    console.log(`\tOutput ${index}`)
    console.log(`\t\tType: ${output.type}`)
    console.log(`\t\tPurpose: ${output.purpose}`)
    console.log(`\t\tAsset: ${output.assetAlias}`)
    console.log(`\t\tAmount: ${output.amount}`)
    console.log(`\t\tAccount: ${output.accountAlias}`)
  })
}
// endsnippet

client.mockHsm.keys.create()
.then(Key => {
  key = Key.xpub
  signer.addKey(key, client.mockHsm.signerConnection)
}).then(() => Promise.all([
  client.assets.create({
    alias: 'gold',
    rootXpubs: [key],
    quorum: 1,
  }),
  client.accounts.create({
    alias: 'alice',
    rootXpubs: [key],
    quorum: 1
  }),
  client.accounts.create({
    alias: 'bob',
    rootXpubs: [key],
    quorum: 1
  }),

  // snippet create-feed
  client.transactionFeeds.create({
    alias: 'local-transactions',
    filter: "is_local='yes'"
  })
  // endsnippet
]))
.then(() => {

  // snippet get-feed
  const feedPromise = client.transactionFeeds.get({
    alias: 'local-transactions'
  })
  // endsnippet

  return feedPromise
})
.then(feed => {

  // snippet processing-loop
  const processingLoop = (tx, next, done, fail) => {
    processTransaction(tx)
    next(true)
  }
  // endsnippet

  // snippet processing-thread
  // JavaScript is single-threaded, and uses a callback for processing
  feed.consume(processingLoop)
  // endsnippet
})
.then(() =>
  // snippet issue
  client.transactions.build(builder => {
    builder.issue({
      assetAlias: 'gold',
      amount: 100
    })
    builder.controlWithAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 100
    })
  })
  .then(issuance => signer.sign(issuance))
  .then(signed => client.transactions.submit(signed))
  // endsnippet
).then(() =>
  // snippet transfer
  client.transactions.build(builder => {
    builder.spendFromAccount({
      accountAlias: 'alice',
      assetAlias: 'gold',
      amount: 50
    })
    builder.controlWithAccount({
      accountAlias: 'bob',
      assetAlias: 'gold',
      amount: 50
    })
  })
  .then(issuance => signer.sign(issuance))
  .then(signed => client.transactions.submit(signed))
  // endsnippet
).catch(err =>
  process.nextTick(() => { throw err })
)
